vultr/resource_vultr_iso_private.go

```diff
@@ -107,7 +109,57 @@ func resourceVultrIsoDelete(ctx context.Context, d *schema.ResourceData, meta in
 	log.Printf("[INFO] Deleting iso : %s", d.Id())
 
 	if err := client.ISO.Delete(ctx, d.Id()); err != nil {
-		return diag.Errorf("error destroying ISO %s : %v", d.Id(), err)
+		// decode the error
+		var attachedErr struct {
+			Error  string `json:"error"`
+			Status int    `json:"status"`
+		}
+
+		if unmarshalError := json.Unmarshal([]byte(err.Error()), &attachedErr); unmarshalError != nil {
+			return diag.Errorf("error deleting ISO %s: parsing error %v in deleting ISO %s : %v", d.Id(), err.Error(), unmarshalError)
+		}
+
+		if !strings.Contains(attachedErr.Error, "is still attached to") {
+			return diag.Errorf("error deleting ISO %s: delete ISO error not related to attachment: delete error %+v", d.Id(), attachedErr)
+		}
+
+		parts := strings.Split(attachedErr.Error, " ")
+		ip := parts[len(parts)-1]
+		if parsedIP := net.ParseIP(ip); parsedIP == nil {
+			return diag.Errorf("error deleting ISO %s : failed to parse IP to which ISO is attached: %s", d.Id(), ip)
+		}
+
+		// default is 100 instances
+		var options govultr.ListOptions
+		for {
+			instances, meta, err := client.Instance.List(ctx, &options)
+			if err != nil {
+				return diag.Errorf("error deleting ISO %s : failed to list instances for detaching ISO", d.Id(), err)
+			}
+
+			// check for the instance with this IP, return on failure or discovery
+			for _, instance := range instances {
+				if instance.MainIP == ip {
+					if err := client.Instance.DetachISO(ctx, instance.ID); err != nil {
+						return diag.Errorf("error deleting ISO %s : failed to detach from instances %s : %s", d.Id(), instance.ID, err)
+					}
+					_, err := waitForIsoDetached(ctx, instance.ID, "ready", []string{"isomounted"}, "status", meta)
+					if err != nil {
+						return diag.Errorf("error deleting ISO %s: failed to wait for ISO to detach from instance %s: %s", d.Id(), instance.ID, err)
+					}
+					if err = client.ISO.Delete(ctx, d.Id()); err != nil {
```

Comment:
Also want to add a `return nil` after this conditional. 


```go
if err = client.ISO.Delete(ctx, d.Id()); err != nil {
	return diag.Errorf("error deleting ISO %s: failed to delete ISO: %s", d.Id(), err)
}
return nil
```

this way we can break out of the for loops since everything has been successfully detached + deleted

Reply:
For sure. Good catch.

Please see https://github.com/vultr/terraform-provider-vultr/pull/131/commits/b8f03f2f328fae9355b64a224f7dd242acbf43f0 .